lib/honeybadger/notice.ex

```diff
@@ -1,43 +1,76 @@
 defmodule Honeybadger.Notice do
   alias Honeybadger.Utils
 
+  @typep error :: %{class: atom | iodata, message: iodata, tags: list, backtrace: list}
+
+  @type noticeable :: Exception.t | Map.t | String.t | atom
+
+  @typep notifier :: %{name: String.t, url: String.t, version: String.t}
+
+  @typep server :: %{environment_name: atom, hostname: String.t, project_root: Path.t}
+
+  @type t :: %__MODULE__{
+    notifier: notifier,
+    server: server,
+    error: error,
+    request: Map.t
+  }
+
+  @enforce_keys [:notifier, :server, :error, :request]
+
   defstruct [:notifier, :server, :error, :request]
 
+  @doc """
+  Create a new `Honeybadger.Notice` struct for various error types.
+
+  ## Example
+
+      iex> Honeybadger.Notice.new("oops", %{}, []).error
+      %{backtrace: [], class: "RuntimeError", message: "oops", tags: []}
 
-  def new(error, metadata \\ %{}, backtrace)
+      iex> Honeybadger.Notice.new(:badarg, %{}, []).error
+      %{backtrace: [], class: "ArgumentError", message: "argument error", tags: []}
+
+      iex> Honeybadger.Notice.new(%RuntimeError{message: "oops"}, %{}, []).error
+      %{backtrace: [], class: "RuntimeError", message: "oops", tags: []}
+  """
+  @spec new(noticeable, Map.t, list) :: t
+  def new(error, metadata, backtrace)
```

Comment:
Do you mind explaining how this line works? I can't remember what Elixir does when there is no function body.

Reply:
It defines a function head without any body. Usually you do this to specify defaults for functions with multiple clauses. In this case I did it for the `@spec` above it. Functions can only have one spec, regardless of how many clauses there are.